Add option to pick-up Monitors cluster-wide (#27)
projectsyn/component-prometheus: docs/modules/ROOT/pages/how-tos/cluster-monitoring.adoc

= Monitor other Commodore Components

The Component Prometheus provides the option to deploy Prometheus and monitor core Kubernetes services.

> However it also supports picking up metrics and alerts from other workloads deployed by other components through the `custer-monitoring` addon and some convenient jsonnet functions.

Reviewer: ```suggestion
However it also supports picking up metrics and alerts from other workloads deployed by other components through the `cluster-monitoring` addon and some convenient jsonnet functions.
```

== Enabling `cluster-monitoring`

To enable monitoring other components we simply need to add the `cluster-monitoring` addon.

> This sets namespace slectors on every Prometheus instance that will result in them picking up all `ServiceMonitors`, `PodMonitors`, `Probes`, and `PrometheusRules` in namespaces with the label `monitoring.syn.tools/<instance>`.

Reviewer: ```suggestion
This sets namespace selectors on every Prometheus instance that will result in them picking up all `ServiceMonitors`, `PodMonitors`, `Probes`, and `PrometheusRules` in namespaces with the label `monitoring.syn.tools/<instance>`.
```

> The example blow will make the Prometheus instance `default-instance` pick up all namespaces with a label `monitoring.syn.tools/default-instance: "true"`.

Reviewer: ```suggestion
The example below will make the Prometheus instance `default-instance` pick up all namespaces with the label `monitoring.syn.tools/default-instance: "true"`.
```

The function `NetworkPolicy` returns a network policy that allows ingress traffic from the Prometheus namespace.
This means when writing a component you don't need to know where Prometheus is deployed.

We also provide helper functions to create `ServiceMonitors`, `PodMonitors`, `Probes`, and `PrometheusRules`.

.Example
[source,jsonnet]
----
  local prometheus = import 'lib/prometheus.libsonnet';

  ...

  local namespace = kube.Namespace(params.namespace)
  {
    '00_namespace': prometheus.RegisterNamespace(namespace),<1>
    '01_networkpolicy': prometheus.NetworkPolicy(){ <2>
      metadata+: {
        namespace: params.namespace,
      },
    },
    '10_servicemonitor': prometheus.ServiceMonitor('foo'){ <3>
      ...
    },
  }
----
<1> Add a label so the default instance will pick up the namespace
<2> Depending on the cluster distribution you will need to add a NetworkPolicy.
Without it Prometheus won't be able to scape the targets.
The `NetworkPolicy` functions will provide a correctly configured NetworkPolicy to allow ingress traffic from the Prometheus instance.
<3> Create a ServiceMonitor called 'foo' that's guaranteed to be picked up by Prometheus.

> WARNING: Don't create a NetworkPolicy for permissive clusters without default NetworkPolicies.

Reviewer: ```suggestion
[WARNING]
Don't create a NetworkPolicy for permissive clusters without default NetworkPolicies.
```

= Monitor other Commodore Components

The Component Prometheus provides the option to deploy Prometheus and monitor core Kubernetes services.

> However it also supports picking up metrics and alerts from other workloads deployed by other components through the `cluster-monitoring` addon and some convenient jsonnet functions.

Reviewer 2: ```suggestion
However it also supports picking up metrics and alerts from other workloads deployed by other components through the `cluster-monitoring` addon and some convenient Jsonnet functions.
```
